[Visualize] New heatmap implementation with elastic-charts

stratoula opened the pull request:
## Summary

Part of https://github.com/elastic/kibana/issues/118516
Part of https://github.com/elastic/kibana/issues/103633
Partially fixes https://github.com/elastic/kibana/issues/30681 (adds a grid)
Closes https://github.com/elastic/kibana/issues/13926
Closes https://github.com/elastic/kibana/issues/10648
Closes https://github.com/elastic/kibana/issues/12666

### PR's output
- Creates a new `visTypeHeatmap` plugin with the new implementation of the visualize heatmap chart. The new implementation is using the elastic-charts library and is **NOT** the default. In order to enable it, you have to switch off the corresponding switch from the advanced settings. 
![image](https://user-images.githubusercontent.com/17003240/141966720-a6a4fe9e-0dfb-464e-895d-994fb833aa0e.png)
- Creates a new `expressionHeatmap` plugin which has the expression function and renderer of the heatmap chart. These are used by both Lens and Visualize editors 🎉 

### Vislib heatmap vs New implementation
- Color scale (is not working correctly in vislib) is not supported
- Rotate and overwrite color values are not supported
- **We always scale to data bounds**. This is super important and the main reason that the users might encounter some differences between the charts (the charts that have the `Scale to data bounds` switch off

For the above, we will expect the user's feedback and decide if we would like to support them in the future

- Small multiples are not yet supported
- Highlight range in hover is not yet supported
- X and Y axis titles are not supported yet
- Heatmap visualization does not show empty buckets with date histogram #100301

The above are blocked from EC side and we would like to support them in order to make the new implementation the default

### Screenshots
dynamic ranges
![image](https://user-images.githubusercontent.com/17003240/141968597-297f1214-82ed-44b7-9811-a446075729a4.png)

percentage mode
![image](https://user-images.githubusercontent.com/17003240/141968869-cac13574-de7a-480c-9ba5-5f83a183dd90.png)

custom ranges and overwrite color
![image](https://user-images.githubusercontent.com/17003240/141969061-2d5ee6a8-d332-4fe3-acdd-beb07fb34b73.png)


### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/6212)
* Commit: 57acd719afffc6f9455dcc6e4b85dffc8cdbe298
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/6212#c94e2194-e26f-49a0-8152-4164c4ff4ac3)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6212","number":6212,"commit":"57acd719afffc6f9455dcc6e4b85dffc8cdbe298"}],"number":6212}
buildkite-pr-comment-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/6230)
* Commit: d2c02e7b8092f68f0197c3780196648e723cce3d
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/6230#1c2b48ed-3286-4c39-b822-b39fd6b3fb86)

### History
* :broken_heart: [Build #6212](https://buildkite.com/elastic/kibana-pull-request/builds/6212) failed 57acd719afffc6f9455dcc6e4b85dffc8cdbe298

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6230","number":6230,"commit":"d2c02e7b8092f68f0197c3780196648e723cce3d"},{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6212","number":6212,"commit":"57acd719afffc6f9455dcc6e4b85dffc8cdbe298"}],"number":6230}
buildkite-pr-comment-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/6247)
* Commit: 981a97e939a7c4af9f9fa46fc61a3f2bdc0cbcca
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/6247#e4a46b31-fe82-4a48-aa49-6847413d764e)

### History
* :broken_heart: [Build #6230](https://buildkite.com/elastic/kibana-pull-request/builds/6230) failed d2c02e7b8092f68f0197c3780196648e723cce3d
* :broken_heart: [Build #6212](https://buildkite.com/elastic/kibana-pull-request/builds/6212) failed 57acd719afffc6f9455dcc6e4b85dffc8cdbe298

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6247","number":6247,"commit":"981a97e939a7c4af9f9fa46fc61a3f2bdc0cbcca"},{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6230","number":6230,"commit":"d2c02e7b8092f68f0197c3780196648e723cce3d"},{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6212","number":6212,"commit":"57acd719afffc6f9455dcc6e4b85dffc8cdbe298"}],"number":6247}
buildkite-pr-comment-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/6281)
* Commit: 446e664f1d200574e6511cd825744af485c3ba1c
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-118338/446e664f1d200574e6511cd825744af485c3ba1c)

### Failed CI Steps
* [Default CI Group #1](https://buildkite.com/elastic/kibana-pull-request/builds/6281#197f5d1b-d6ee-4f95-b7fd-fdf67031d36c)
* [Default CI Group #16](https://buildkite.com/elastic/kibana-pull-request/builds/6281#8acce4b0-e060-4f19-b8fd-3eaed13c218b)
* [OSS Misc Functional Tests](https://buildkite.com/elastic/kibana-pull-request/builds/6281#4699e0cc-27bd-4100-a8a5-d6293f819205)
* [Jest Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/6281#8732aa75-685a-4215-bfb8-e1f2af4c8712)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/6281#8732aa75-685a-4215-bfb8-e1f2af4c8712) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/6281/jobs/8732aa75-685a-4215-bfb8-e1f2af4c8712/artifacts/dbf62bc2-b83f-420b-870d-2c439e049c18) Jest Tests #<span></span>8 / interpreter/functions#heatmap logs correct datatable to inspector
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/6281#8732aa75-685a-4215-bfb8-e1f2af4c8712) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/6281/jobs/8732aa75-685a-4215-bfb8-e1f2af4c8712/artifacts/86faaee4-d0c8-436d-9229-72a1c481a5d3) Jest Tests #<span></span>8 / interpreter/functions#heatmap returns an object with the correct structure
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/6281#8acce4b0-e060-4f19-b8fd-3eaed13c218b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/6281/jobs/8acce4b0-e060-4f19-b8fd-3eaed13c218b/artifacts/f1349e59-cb0a-4498-ad05-ff7cbc963689) Default CI Group #<span></span>16 / lens app  lens add-to-dashboards tests should add a Lens heatmap to the dashboard
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/6281#197f5d1b-d6ee-4f95-b7fd-fdf67031d36c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/6281/jobs/197f5d1b-d6ee-4f95-b7fd-fdf67031d36c/artifacts/021d4a08-f39d-43f6-8344-dd73d7466916) Default CI Group #<span></span>1 / Stack Management registers all UI Settings in the UsageStats interface
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/6281#4699e0cc-27bd-4100-a8a5-d6293f819205) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/6281/jobs/4699e0cc-27bd-4100-a8a5-d6293f819205/artifacts/2efc554d-cf9d-461f-9093-9967aae910c5) OSS Misc Functional Tests / telemetry Telemetry service detects that telemetry cannot be sent in screenshot mode

### History
* :broken_heart: [Build #6247](https://buildkite.com/elastic/kibana-pull-request/builds/6247) failed 981a97e939a7c4af9f9fa46fc61a3f2bdc0cbcca
* :broken_heart: [Build #6230](https://buildkite.com/elastic/kibana-pull-request/builds/6230) failed d2c02e7b8092f68f0197c3780196648e723cce3d
* :broken_heart: [Build #6212](https://buildkite.com/elastic/kibana-pull-request/builds/6212) failed 57acd719afffc6f9455dcc6e4b85dffc8cdbe298

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6281","number":6281,"commit":"446e664f1d200574e6511cd825744af485c3ba1c"},{"buildStatus":{"state":"running","success":false,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6247","number":6247,"commit":"981a97e939a7c4af9f9fa46fc61a3f2bdc0cbcca"},{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6230","number":6230,"commit":"d2c02e7b8092f68f0197c3780196648e723cce3d"},{"buildStatus":{"state":"running","success":false,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/6212","number":6212,"commit":"57acd719afffc6f9455dcc6e4b85dffc8cdbe298"}],"number":6281}
buildkite-pr-comment-->

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

stratoula commented on src/plugins/chart_expressions/expression_heatmap/common/expression_functions/heatmap_function.ts:
```diff
@@ -0,0 +1,201 @@
+  label: string
+) => {
+  const dimension =
+    typeof accessor === 'string' ? convertToVisDimension(columns, accessor) : accessor;
```
ℹ️ The new inspector adapter accepts `vis_dimension` as argument

stratoula commented on src/plugins/chart_expressions/expression_heatmap/common/expression_functions/heatmap_function.ts:
```diff
@@ -0,0 +1,201 @@
+        defaultMessage: 'Configure the heatmap layout.',
+      }),
+    },
+    showTooltip: {
```
ℹ️ This setting is only supported by Viz editor for now, but could be considered as an addition for Lens. There are users that don't want to have tooltips.

stratoula commented on src/plugins/chart_expressions/expression_heatmap/common/expression_functions/heatmap_function.ts:
```diff
@@ -0,0 +1,201 @@
+      help: i18n.translate('expressionHeatmap.function.args.highlightInHoverHelpText', {
+        defaultMessage:
+          'When this is enabled, it highlights the ranges of the same color on legend hover',
+      }),
```
ℹ️ Supported by vislib heatmap for now, but is something that we would like to also support for the new implementation. Blocked by EC for now. We can also consider it for Lens (maybe not a setting but as a default behavior)

stratoula commented on src/plugins/chart_expressions/expression_heatmap/common/expression_functions/heatmap_function.ts:
```diff
@@ -0,0 +1,201 @@
+        defaultMessage: 'If is set to true, the last range value will be right open',
+      }),
+      default: true,
+    },
```
❓To be discussed with the team. The purpose of this flag is to change the last range format. For Lens it is open
![image](https://user-images.githubusercontent.com/17003240/141959400-2e0e86d6-5b54-46da-9f28-94b7b4d01d90.png)

while in Viz editor is set to the maximum value
![image](https://user-images.githubusercontent.com/17003240/141959455-e6b3d1e1-3d25-4e1b-82d3-8bcf0e9a7edb.png)

Two options here:
- Support it as it is for a smoother transition from vislib (I am open to change the name, I couldn't think of a better name)
- Not support it and follow the Lens way (last range will be set to Infinite).

I don't have a strong opinion! Let's decide together :)

stratoula commented on src/plugins/chart_expressions/expression_heatmap/public/utils/get_color_picker.tsx:
```diff
@@ -0,0 +1,85 @@
+          closePopover={onClose}
+          panelPaddingSize="s"
+        >
+          <ColorPicker
```
ℹ️ I am using the EUI colors for the picker here and not the vislib ones. As the vislib ones are considered deprecated I think is a good opportunity to use them here instead.

<img width="345" alt="Screenshot 2021-11-16 at 11 38 27 AM" src="https://user-images.githubusercontent.com/17003240/141960457-69a8a53f-cd31-4056-9958-3af44405c87f.png">

stratoula commented on src/plugins/visualize/public/application/components/visualize_editor_common.tsx:
```diff
@@ -126,6 +131,7 @@ export const VisualizeEditorCommon = ({
         />
       )}
       {visInstance?.vis?.type?.stage === 'experimental' && <ExperimentalVisInfo />}
+      {!hasHeatmapLegacyhartsEnabled && isSplitChart && <SplitChartWarning />}
```
ℹ️ Displays a notification (visible only from the editor) for the heatmap charts with the new implementation and the split chart bucket set to column or row. This appears only for already existing charts (I dont allow the split chart for new charts).
![image](https://user-images.githubusercontent.com/17003240/141961995-0bd0b40e-2713-4533-a65f-b9aa704659cc.png)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

jbudz reviewed (approved):
limits.yml LGTM

dej611 commented on src/plugins/chart_expressions/expression_heatmap/common/expression_functions/heatmap_legend.ts:
```diff
@@ -1,84 +1,62 @@
+import { EXPRESSION_HEATMAP_LEGEND_NAME } from '../constants';
+import { HeatmapLegendConfig, HeatmapLegendConfigResult } from '../types';
 
 /**
  * TODO check if it's possible to make a shared function
  * based on the XY chart
  */
```
I think this can be removed

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/empty_placeholder.tsx:
```diff
@@ -0,0 +1,23 @@
+/*
```
There's a typo in the folder name: `compoments` => `components`

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.test.tsx:
```diff
@@ -0,0 +1,247 @@
+import { chartPluginMock } from '../../../../charts/public/mocks';
+import { fieldFormatsServiceMock } from '../../../../field_formats/public/mocks';
+import type { Datatable } from '../../../../expressions/public';
+// import { shallow, mount } from 'enzyme';
```
remove

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.test.tsx:
```diff
@@ -0,0 +1,247 @@
+  setSilent: jest.fn(),
+} as any;
+
+describe('PieComponent', function () {
```
`HeatmapComponent`

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.test.tsx:
```diff
@@ -0,0 +1,247 @@
+    await act(async () => {
+      expect(component.find(Heatmap).prop('colorScale')).toEqual({
+        bands: [
+          { color: 'rgb(0, 0, 0, 0)', end: 0, start: 0 },
```
Isn't this `rgb` string wrong? It think it should be  either`rgb(0,0,0)` or `rgba(0,0,0,0)`

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.tsx:
```diff
@@ -113,7 +112,20 @@ function computeColorRanges(
 
-export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+const getAccessor = (value: string | ExpressionValueVisDimension, columns: DatatableColumn[]) => {
+  if (typeof value === 'string') {
+    return value;
+  } else {
+    const accessor = value.accessor;
+    if (typeof accessor === 'number') {
+      return columns[accessor].id;
+    } else {
+      return accessor.id;
+    }
+  }
+};
```
nitpick: this can be flattened as there's a `return` statement in each branch
```suggestion
  if (typeof value === 'string') {
    return value;
  }
  const accessor = value.accessor;
  if (typeof accessor === 'number') {
    return columns[accessor].id;
  }
  return accessor.id;
};
```

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.tsx:
```diff
@@ -122,33 +134,74 @@ export const HeatmapComponent: FC<HeatmapRenderProps> = ({
   const isDarkTheme = chartsThemeService.useDarkMode();
+  // legacy heatmap legend is handled by the uiState
+  const [showLegend, setShowLegend] = useState<boolean>(() => {
+    const bwcLegendStateDefault = args.legend.isVisible == null ? true : args.legend.isVisible;
```
nitpick:

```suggestion
    const bwcLegendStateDefault = args.legend.isVisible ?? true;
```

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.tsx:
```diff
@@ -159,17 +212,23 @@ export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+  const xAxisMeta = xAxisColumn?.meta;
+  const xValuesFormatter = formatFactory(xAxisMeta?.params);
+  const metricFormatter = formatFactory(
+    typeof args.valueAccessor === 'string' ? valueColumn.meta.params : args?.valueAccessor?.format
```
nitpick:

```suggestion
    typeof valueAccessor === 'string' ? valueColumn.meta.params : args?.valueAccessor?.format
```

stratoula replied:
I need the args.valueAceessor here to make this check because it can be either a string or a vis_dimension. In case of vis_dimension I am taking the meta params from it.

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/compoments/heatmap_component.tsx:
```diff
@@ -349,56 +432,65 @@ export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+          onBrushEnd={onBrushEnd as BrushEndListener}
+        />
+        <Heatmap
+          id="heatmap"
```
why removing the `tableId` here?
Can we do `id={tableId ?? 'heatmap'}`?

stratoula replied:
We have only one table now (we dont use the multitable so we cant do this!

dej611 replied:
I was looking at the `id` type description and it mentions that it should be unique:
```ts
export interface Spec {
    /** unique Spec identifier */
    id: string;
    /** Chart type define the type of chart that use this spec */
    chartType: ChartType;
    /** The type of spec, can be series, axis, annotation, settings etc */
    specType: string;
}
```

I've tested in dashboard and see no problem- I was assuming it was assigning the `id` to the canvas element. Maybe somebody from @elastic/datavis team can clarify the uniqueness constraint in this case?

stratoula replied:
I discussed it offline with MarcoV and made it unique!

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/utils/get_color_picker.tsx:
```diff
@@ -0,0 +1,85 @@
+  ({ anchor, color, onClose, onChange, seriesIdentifiers: [seriesIdentifier] }) => {
+    const seriesName = seriesIdentifier.key;
+    const overwriteColors: Record<string, string> = uiState?.get('vis.colors', {}) ?? {};
+    const colorIsOverwritten = Object.keys(overwriteColors).includes(seriesName.toString());
```
```suggestion
    const colorIsOverwritten = seriesName.toString() in overwriteColors;
```

dej611 commented on src/plugins/chart_expressions/expression_heatmap/public/utils/get_timezone.ts:
```diff
@@ -0,0 +1,23 @@
+export function getTimeZone(uiSettings: IUiSettingsClient) {
+  if (uiSettings.isDefault('dateFormat:tz')) {
+    const detectedTimeZone = moment.tz.guess();
+    if (detectedTimeZone) return detectedTimeZone;
+    else return moment().format('Z');
```
```suggestion
    return detectedTimeZone || moment().format('Z');
```

dej611 commented on src/plugins/visualize/public/application/components/split_chart_warning.tsx:
```diff
@@ -0,0 +1,66 @@
+                {!canEditAdvancedSettings && (
+                  <FormattedMessage
+                    id="visualize.legacyCharts.conditionalMessage.noPermissions"
+                    defaultMessage="Contact your system administrator to switch to the new library."
```
Shouldn't this be:

```suggestion
                    defaultMessage="Contact your system administrator to switch to the old library."
```
?

stratoula replied:
Correct :D

dej611 reviewed (commented):
Had a first code review (only). Left some feedback.
Will test locally next

stratoula commented:
@elasticmachine merge upstream

mbondyra commented on src/plugins/chart_expressions/expression_heatmap/public/constants.ts:
```diff
@@ -0,0 +1,28 @@
+export const DEFAULT_MIN_STOP = 0;
+export const DEFAULT_MAX_STOP = 100;
+export const DEFAULT_COLOR_STEPS = 5;
+export const DEFAULT_COLOR = '#6092C0'; // Same as EUI ColorStops default for new stops
```
nit: maybe it makes sense to import it the way it is imported here? 
https://github.com/elastic/kibana/blob/main/x-pack/plugins/ml/common/util/group_color_utils.ts#L14

stratoula replied:
Well it is not used so I will delete it :D

mbondyra commented:
When testing, I noticed that the `reset colors` link-button is gone from options panel from Heatmap settings section:
<img width="355" alt="Screenshot 2021-11-17 at 17 18 19" src="https://user-images.githubusercontent.com/4283304/142239016-e5360d23-3286-4389-b56d-106bf73a74a5.png">

dej611 reviewed (changes requested):
Testing locally (Safari and Chrome):

* When brushing in Visualize ( even the starting heatmap chart ) nothing happens, and the following error appears in console. No error in Lens instead.
```
Uncaught TypeError: Cannot read properties of undefined (reading 'meta')
    at onBrushEnd (heatmap_component.tsx:300)
    at on_brush_end_caller.js:32
    at resultFuncWithRecomputations (index.js:77)
    at index.js:70
    at index.js:30
    at index.js:84
    at index.js:30
    at selector (index.js:105)
    at HeatmapState.onBrushEndCaller (on_brush_end_caller.js:38)
    at HeatmapState.eventCallbacks (chart_state.js:99)
```

* The Color scale input seems disabled with the new library but no tooltip is shown as with other inputs

* X Axis title is not shown with the new version. Same happens for Metric/Y Axis.

Old library:
<img width="1436" alt="Screenshot 2021-11-17 at 17 35 37" src="https://user-images.githubusercontent.com/924948/142242507-de57e2e9-9a96-446b-9de3-e0d3d841ccc1.png">

New library:
<img width="1440" alt="Screenshot 2021-11-17 at 17 36 32" src="https://user-images.githubusercontent.com/924948/142242560-0d2c822f-19f3-4e7d-a1f7-402a4db0ce93.png">

I think 1 and 3 requires to be addressed. As for 2 it would be nice to have it as well, but it's a minor thing I'd say.

stratoula commented on src/plugins/chart_expressions/expression_heatmap/public/components/empty_placeholder.tsx:
```diff
@@ -0,0 +1,23 @@
+import { EuiIcon, EuiText, IconType, EuiSpacer } from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n/react';
+
+export const EmptyPlaceholder = (props: { icon: IconType }) => (
```
Discussed with Marta offline. I will move this to the charts plugin

stratoula commented:
@dej611 thanx for the bug! The xy titles are not yet supported from EC side :)

MichaelMarcialis commented on src/plugins/charts/public/static/components/empty_placeholder.tsx:
```diff
@@ -0,0 +1,23 @@
+import { FormattedMessage } from '@kbn/i18n/react';
+
+export const EmptyPlaceholder = (props: { icon: IconType }) => (
+  <>
+    <EuiText className="heatmap-chart__empty" textAlign="center" color="subdued" size="xs">
+      <EuiIcon type={props.icon} color="subdued" size="l" />
+      <EuiSpacer size="s" />
+      <p>
+        <FormattedMessage id="charts.noDataLabel" defaultMessage="No results found" />
+      </p>
+    </EuiText>
+  </>
```
Would it be better to utilize an `EuiEmptyPrompt` component here? Or would that deviate from how such messages are delivered elsewhere in Visualize?

stratoula replied:
I haven't made any change on this component. It is the component that is used in Lens, I just moved it to the charts plugin in order to be accessible by other plugins
![image](https://user-images.githubusercontent.com/17003240/142992838-c7d753d3-32cb-4dad-b497-4a5a45bd7de0.png)
If I use `EuiEmptyPrompt ` this will also affect Lens (not only Visualize) so it will be better to not do this in this PR. Wdyt?

MichaelMarcialis replied:
Yeah, if this pattern is used elsewhere, let's keep it as is for now. We can re-evaluate in a more holistic fashion later.

MichaelMarcialis commented on src/plugins/vis_types/heatmap/public/vis_type/heatmap.tsx:
```diff
@@ -0,0 +1,129 @@
+      enableHover: false,
+      legendPosition: Position.Right,
+      colorsNumber: 4,
+      colorSchema: ColorSchemas.Greens,
```
Here the heatmap visualizations default to using the "Greens" (aka positive) color palette. However, in Lens, the heatmap visualization defaults to using the temperature color palette. Currently it doesn't look like Visualize offers the option for a temperature color palette. If we are unable or unwilling to add the temperature color palette and apply it as the default for consistency, should we change default palette to be "Green to Red" (aka status) to make it a little closer conceptually?

stratoula replied:
@MichaelMarcialis we are going to use the EUI palettes on the heatmap on a next iteration and then we will add the temperature palette as the default https://github.com/elastic/kibana/issues/116917
For now, I will follow your suggestions and all the new heatmap visualizations will default to the GreenToRed color schema.

MichaelMarcialis reviewed (commented):
Hey, @stratoula. This is looking great so far! I'm leaving you a few comments and questions below for your review. Thanks!

- Is it typical for Kibana to have deprecated advanced settings that are enabled by default? I was just curious, as it felt odd to not default to the new elastic charts heatmap (if the previous implementation has been officially deprecated). Alternatively, we could also flip it so that the new heatmap implementation is a beta/experimental feature that is off by default (that makes more sense in my head). Thoughts?

- I believe we also have this issue in Lens, but I noticed that lengthy y-axis labels can cause the available horizontal space for the visualization to be extremely small. Do we have any ability to truncate or break lines for these labels?

![image](https://user-images.githubusercontent.com/3884767/142469396-e9354e8e-60ee-45a4-af25-3d44978d4d70.png)

- There doesn't appear to be any whitespace around the edges of the visualization. This has the potential to cause collisions between the visualization and the bottom-left legend toggle button. Is it possible to bake in some whitespace into the visualization to avoid such collisions?

![image](https://user-images.githubusercontent.com/3884767/142469739-d4474a03-060e-41bf-883c-c9e7934011bf.png)

stratoula commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

stratoula commented:
> Hey, @stratoula. This is looking great so far! I'm leaving you a few comments and questions below for your review. Thanks!
> 
> * Is it typical for Kibana to have deprecated advanced settings that are enabled by default? I was just curious, as it felt odd to not default to the new elastic charts heatmap (if the previous implementation has been officially deprecated). Alternatively, we could also flip it so that the new heatmap implementation is a beta/experimental feature that is off by default (that makes more sense in my head). Thoughts?
>
You are right! I removed the deprecated message. This should be deprecated when the new implementation is the default.

> * I believe we also have this issue in Lens, but I noticed that lengthy y-axis labels can cause the available horizontal space for the visualization to be extremely small. Do we have any ability to truncate or break lines for these labels?
> 
Yes this also exists in Lens. @markov00 is this possible? I would suggest if this is possible to address it in different PR though.
> ![image](https://user-images.githubusercontent.com/3884767/142469396-e9354e8e-60ee-45a4-af25-3d44978d4d70.png)
> 
> * There doesn't appear to be any whitespace around the edges of the visualization. This has the potential to cause collisions between the visualization and the bottom-left legend toggle button. Is it possible to bake in some whitespace into the visualization to avoid such collisions?
> 
> ![image](https://user-images.githubusercontent.com/3884767/142469739-d4474a03-060e-41bf-883c-c9e7934011bf.png)

I added more padding and I think it looks better now. A note here: One thing missing are the axis titles. When they will be supported it will give more space to the axis and we won't have this kind of problems.

stratoula commented:
@elasticmachine merge upstream

mbondyra reviewed (approved):
Retested thoroughly and couldn't find anything wrong! Code LGTM 👌🏼

mbondyra commented on src/plugins/chart_expressions/expression_heatmap/common/types/expression_functions.ts:
```diff
@@ -0,0 +1,101 @@
+import { ExpressionValueVisDimension } from '../../../../visualizations/common';
+
+import { CustomPaletteState, PaletteOutput } from '../../../../charts/common';
+// import { VisParams, visType } from './expression_renderers';
```
this can be removed, right?

stratoula replied:
Yesss!

dej611 reviewed (approved):
Tested in Safari and all works fine 👍

stratoula commented:
@MichaelMarcialis are you ok with the current status? I need your approval to merge :D

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@elasticmachine merge upstream

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @stratoula. I have one last question below, but nothing worth holding you up over. Approving.

~The question I have is regarding the legend toggle button. With the new padding you've added to the visualization container, should we remove the padding to the `.visLegend__toggle` button in order to reduce the chance of overlapping with the visualization and to keep it's position consistent with other legend toggle buttons (such as those in XY charts)?~

stratoula commented:
Discussed offline with Michael, just adding it here for visibility. The `.visLegend__toggle` is used by the vislib heatmap, vislib pie and gauge/goal. I have made no changes to the vislib component so we could remove the padding from the toggle legend button but I suggest to do it in another PR as is irrelevant with the changes we made here

MichaelMarcialis commented:
@stratoula: Ignore my question above. I had forgotten to disable the legacy heatmaps in advanced settings. What you have here is good. Thanks again!
